server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/NodeShutdownAllocationDeciderTests.java

```diff
@@ -157,43 +151,103 @@ public void testCanAutoExpandToNodeThatIsNotShuttingDown() {
-            RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, state, null, null, 0);
-            allocation.debugDecision(true);
+        for (SingleNodeShutdownMetadata.Type type : List.of(
+            SingleNodeShutdownMetadata.Type.REMOVE,
+            SingleNodeShutdownMetadata.Type.SIGTERM
+        )) {
+            ClusterState state = prepareState(type);
+            RoutingAllocation allocation = createRoutingAllocation(state);
 
             Decision decision = decider.shouldAutoExpandToNode(indexMetadata, DATA_NODE, allocation);
             assertThat(decision.type(), equalTo(Decision.Type.NO));
             assertThat(decision.getExplanation(), equalTo("node [" + DATA_NODE.getId() + "] is preparing to be removed from the cluster"));
         }
     }
 
-    private ClusterState prepareState(ClusterState initialState, SingleNodeShutdownMetadata.Type shutdownType) {
-        return prepareState(initialState, shutdownType, DATA_NODE.getId());
-    }
+    public void testAutoExpandDuringNodeReplacement() {
 
-    private ClusterState prepareState(ClusterState initialState, SingleNodeShutdownMetadata.Type shutdownType, String nodeId) {
-        final String targetNodeName = shutdownType == SingleNodeShutdownMetadata.Type.REPLACE ? randomAlphaOfLengthBetween(10, 20) : null;
-        final SingleNodeShutdownMetadata nodeShutdownMetadata = SingleNodeShutdownMetadata.builder()
-            .setNodeId(nodeId)
-            .setType(shutdownType)
-            .setReason(this.getTestName())
-            .setStartedAtMillis(1L)
-            .setTargetNodeName(targetNodeName)
-            .setGracePeriod(
-                shutdownType == SingleNodeShutdownMetadata.Type.SIGTERM
-                    ? TimeValue.parseTimeValue(randomTimeValue(), this.getTestName())
-                    : null
-            )
+        var state = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(DiscoveryNodes.builder().add(DATA_NODE).build())
+            .metadata(Metadata.builder().put(IndexMetadata.builder(indexMetadata)))
             .build();
-        NodesShutdownMetadata nodesShutdownMetadata = new NodesShutdownMetadata(new HashMap<>()).putSingleNodeMetadata(
-            nodeShutdownMetadata
+        ;
+
+        // should auto-expand when no shutdown
+        assertThatDecision(
+            decider.shouldAutoExpandToNode(indexMetadata, DATA_NODE, createRoutingAllocation(state)),
+            Decision.Type.YES,
+            "no nodes are shutting down"
+        );
+
+        // should auto-expand when shutdown/replacement is registered and no node replacement yet
+        var shutdown = createNodesShutdownMetadata(SingleNodeShutdownMetadata.Type.REPLACE, DATA_NODE.getId());
+        state = ClusterState.builder(state)
+            .metadata(Metadata.builder(state.metadata()).putCustom(NodesShutdownMetadata.TYPE, shutdown).build())
+            .build();
+        assertThatDecision(
```

Comment:
Can we add a verification that a shard that is not yet on the source node cannot expand to it?

Reply:
I did not find this comment addressed?

Reply:
NodeShutdownAllocationDeciderTests only prevents allocations and auto expands on the node that is about to be shut down and do not restrict allocations on the replacement node.

A scenario `shard from any node in cluster can not be auto-expanded on removing node` is tested in: `testCannotAutoExpandToRemovingNode`
A similar one for replacing node is tested in: testAutoExpandDuringNodeReplacement. Please note no ShardRouting created as both tests do not depend on the source shard location (from NodeShutdownAllocationDecider point of view).

Reply:
Sorry for the confusion, I meant this for the replace decider. See [this comment](https://github.com/elastic/elasticsearch/pull/96281/files#r1219134129).